diff --git a/Libraries/Components/Keyboard/KeyboardAvoidingView.js b/Libraries/Components/Keyboard/KeyboardAvoidingView.js
index 6eca1569f3024881ffd645aa026dc5ee4090d93b..5a42107172b8a5c9bd26b89b8207377e15ee9125 100644
--- a/Libraries/Components/Keyboard/KeyboardAvoidingView.js
+++ b/Libraries/Components/Keyboard/KeyboardAvoidingView.js
@@ -208,7 +208,7 @@ class KeyboardAvoidingView extends React.Component<
       ];
     } else {
       this._subscriptions = [
-        Keyboard.addListener('keyboardDidHide', this._onKeyboardChange),
+        Keyboard.addListener('keyboardDidHide', this._onKeyboardHide),
         Keyboard.addListener('keyboardDidShow', this._onKeyboardChange),
       ];
     }
diff --git a/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java b/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java
index 6c625367010a9a3a9bb3c8664a945312129ff3ab..65b9e05a051851c0d4045cabad73aff8b7a88538 100644
--- a/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java
+++ b/ReactAndroid/src/main/java/com/facebook/react/ReactRootView.java
@@ -16,7 +16,6 @@ import static com.facebook.systrace.Systrace.TRACE_TAG_REACT;
 import android.content.Context;
 import android.graphics.BlendMode;
 import android.graphics.Canvas;
-import android.graphics.Insets;
 import android.graphics.Paint;
 import android.graphics.Point;
 import android.graphics.Rect;
@@ -24,18 +23,18 @@ import android.os.Build;
 import android.os.Bundle;
 import android.util.AttributeSet;
 import android.util.DisplayMetrics;
-import android.view.DisplayCutout;
 import android.view.KeyEvent;
 import android.view.MotionEvent;
 import android.view.Surface;
 import android.view.View;
 import android.view.ViewGroup;
 import android.view.ViewTreeObserver;
-import android.view.WindowInsets;
 import android.view.WindowManager;
 import android.widget.FrameLayout;
 import androidx.annotation.Nullable;
-import androidx.annotation.RequiresApi;
+import androidx.core.graphics.Insets;
+import androidx.core.view.ViewCompat;
+import androidx.core.view.WindowInsetsCompat;
 import com.facebook.common.logging.FLog;
 import com.facebook.infer.annotation.Assertions;
 import com.facebook.infer.annotation.ThreadConfined;
@@ -730,11 +729,7 @@ public class ReactRootView extends FrameLayout implements RootView, ReactRoot {
 
   @VisibleForTesting
   /* package */ void simulateCheckForKeyboardForTesting() {
-    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-      getCustomGlobalLayoutListener().checkForKeyboardEvents();
-    } else {
-      getCustomGlobalLayoutListener().checkForKeyboardEventsLegacy();
-    }
+    getCustomGlobalLayoutListener().checkForKeyboardEvents();
   }
 
   private CustomGlobalLayoutListener getCustomGlobalLayoutListener() {
@@ -858,16 +853,13 @@ public class ReactRootView extends FrameLayout implements RootView, ReactRoot {
 
   private class CustomGlobalLayoutListener implements ViewTreeObserver.OnGlobalLayoutListener {
     private final Rect mVisibleViewArea;
-    private final int mMinKeyboardHeightDetected;
 
     private boolean mKeyboardIsVisible = false;
-    private int mKeyboardHeight = 0; // Only used in checkForKeyboardEventsLegacy path
     private int mDeviceRotation = 0;
 
     /* package */ CustomGlobalLayoutListener() {
       DisplayMetricsHolder.initDisplayMetricsIfNotInitialized(getContext().getApplicationContext());
       mVisibleViewArea = new Rect();
-      mMinKeyboardHeightDetected = (int) PixelUtil.toPixelFromDIP(60);
     }
 
     @Override
@@ -876,31 +868,25 @@ public class ReactRootView extends FrameLayout implements RootView, ReactRoot {
         return;
       }
 
-      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) {
-        checkForKeyboardEvents();
-      } else {
-        checkForKeyboardEventsLegacy();
-      }
-
+      checkForKeyboardEvents();
       checkForDeviceOrientationChanges();
       checkForDeviceDimensionsChanges();
     }
 
-    @RequiresApi(api = Build.VERSION_CODES.R)
     private void checkForKeyboardEvents() {
       getRootView().getWindowVisibleDisplayFrame(mVisibleViewArea);
-      WindowInsets rootInsets = getRootView().getRootWindowInsets();
+      WindowInsetsCompat rootInsets = ViewCompat.getRootWindowInsets(getRootView());
       if (rootInsets == null) {
         return;
       }
 
-      boolean keyboardIsVisible = rootInsets.isVisible(WindowInsets.Type.ime());
+      boolean keyboardIsVisible = rootInsets.isVisible(WindowInsetsCompat.Type.ime());
       if (keyboardIsVisible != mKeyboardIsVisible) {
         mKeyboardIsVisible = keyboardIsVisible;
+        Insets barInsets = rootInsets.getInsets(WindowInsetsCompat.Type.systemBars());
 
         if (keyboardIsVisible) {
-          Insets imeInsets = rootInsets.getInsets(WindowInsets.Type.ime());
-          Insets barInsets = rootInsets.getInsets(WindowInsets.Type.systemBars());
+          Insets imeInsets = rootInsets.getInsets(WindowInsetsCompat.Type.ime());
           int height = imeInsets.bottom - barInsets.bottom;
 
           ViewGroup.LayoutParams rootLayoutParams = getRootView().getLayoutParams();
@@ -923,7 +909,7 @@ public class ReactRootView extends FrameLayout implements RootView, ReactRoot {
           sendEvent(
               "keyboardDidHide",
               createKeyboardEventPayload(
-                  PixelUtil.toDIPFromPixel(mVisibleViewArea.height()),
+                  PixelUtil.toDIPFromPixel(mVisibleViewArea.bottom + barInsets.bottom),
                   0,
                   PixelUtil.toDIPFromPixel(mVisibleViewArea.width()),
                   0));
@@ -931,54 +917,6 @@ public class ReactRootView extends FrameLayout implements RootView, ReactRoot {
       }
     }
 
-    private void checkForKeyboardEventsLegacy() {
-      getRootView().getWindowVisibleDisplayFrame(mVisibleViewArea);
-
-      int notchHeight = 0;
-      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
-        WindowInsets insets = getRootView().getRootWindowInsets();
-        if (insets != null) {
-          DisplayCutout displayCutout = insets.getDisplayCutout();
-          if (displayCutout != null) {
-            notchHeight = displayCutout.getSafeInsetTop();
-          }
-        }
-      }
-      final int heightDiff =
-          DisplayMetricsHolder.getWindowDisplayMetrics().heightPixels
-              - mVisibleViewArea.bottom
-              + notchHeight;
-
-      boolean isKeyboardShowingOrKeyboardHeightChanged =
-          mKeyboardHeight != heightDiff && heightDiff > mMinKeyboardHeightDetected;
-
-      if (isKeyboardShowingOrKeyboardHeightChanged) {
-        mKeyboardHeight = heightDiff;
-        mKeyboardIsVisible = true;
-        sendEvent(
-            "keyboardDidShow",
-            createKeyboardEventPayload(
-                PixelUtil.toDIPFromPixel(mVisibleViewArea.bottom),
-                PixelUtil.toDIPFromPixel(mVisibleViewArea.left),
-                PixelUtil.toDIPFromPixel(mVisibleViewArea.width()),
-                PixelUtil.toDIPFromPixel(mKeyboardHeight)));
-        return;
-      }
-
-      boolean isKeyboardHidden = mKeyboardHeight != 0 && heightDiff <= mMinKeyboardHeightDetected;
-      if (isKeyboardHidden) {
-        mKeyboardHeight = 0;
-        mKeyboardIsVisible = false;
-        sendEvent(
-            "keyboardDidHide",
-            createKeyboardEventPayload(
-                PixelUtil.toDIPFromPixel(mVisibleViewArea.height()),
-                0,
-                PixelUtil.toDIPFromPixel(mVisibleViewArea.width()),
-                0));
-      }
-    }
-
     private void checkForDeviceOrientationChanges() {
       final int rotation =
           ((WindowManager) getContext().getSystemService(Context.WINDOW_SERVICE))
